src/shipPlacer.ts

```diff
@@ -0,0 +1,171 @@
+import {ShipPossibilities} from './shipPossibilities'
+import {ShotImpliedPlacements} from './shotImpliedPlacements'
+import * as Globals from './constants'
+import {ShipPlacement} from './shipPlacement'
+import {Position} from './position'
+import {Direction} from './direction'
+import {Shot} from './shot'
+
+export class ShipPlacer{
+    private currentWeightsByLength: {[length:number]: Array<Array<Array<{horizontal: number, vertical: number}>>>} = {};
+    private dynamicWeightsByLength: {[length:number]: Array<Array<{horizontal: number, vertical: number}>>} = {};
+    private shipPossibilitiesByLength: {[length:number]: ShipPossibilities} = {};
+    public shipLengths: Array<number> = [];
+    public numberOfShipsOfLength: {[length: number]: number} = {};
+
+    public constructor(shipLengths:Array<number>){
+        this.shipLengths = shipLengths;
+
+        for(var i=0; i<this.shipLengths.length; i++){
+            if(!this.numberOfShipsOfLength[shipLengths[i]]){
+                this.numberOfShipsOfLength[shipLengths[i]] = 1;
+            }else{
+                this.numberOfShipsOfLength[shipLengths[i]]++;
+            }
+        }
+
+        this.remakePossibilities();
+
+        this.zeroWeights();
+        
+        this.initializeDynamicWeights();
+
+        this.setInvalidPositionsToNegInf();
+
+        this.addDynamicWeights()
+        
+    }
```

Comment:
(Though you could probably cut every line of whitespace following `this.remakePossibilities();`